src/compiler/binder.rs

```diff
@@ -63,19 +71,113 @@ impl BinderType {
         *self.parent.try_write().unwrap() = parent;
     }
 
+    fn container(&self) -> Rc<Node> {
+        self.container.try_read().unwrap().as_ref().unwrap().clone()
+    }
+
+    fn maybe_container(&self) -> Option<Rc<Node>> {
+        self.container
+            .try_read()
+            .unwrap()
+            .as_ref()
+            .map(Clone::clone)
+    }
+
+    fn set_container(&self, container: Option<Rc<Node>>) {
+        *self.container.try_write().unwrap() = container;
+    }
+
+    #[allow(non_snake_case)]
+    fn Symbol(&self) -> fn(SymbolFlags, __String) -> Symbol {
+        self.Symbol.try_read().unwrap().unwrap()
+    }
+
+    #[allow(non_snake_case)]
+    fn set_Symbol(&self, Symbol: fn(SymbolFlags, __String) -> Symbol) {
+        *self.Symbol.try_write().unwrap() = Some(Symbol);
+    }
+
     fn bind_source_file(&self, f: Rc<Node>) {
         self.set_file(Some(f.clone()));
 
+        self.set_Symbol(object_allocator.get_symbol_constructor());
+
         if true {
             self.bind(Some(self.file()));
         }
 
         self.set_file(None);
         self.set_parent(None);
+        self.set_container(None);
+    }
+
+    fn create_symbol(&self, flags: SymbolFlags, name: __String) -> Symbol {
+        self.Symbol()(flags, name)
+    }
+
+    fn add_declaration_to_symbol(&self, symbol: Rc<Symbol>, node: Rc<Node /*Declaration*/>) {
+        node.set_symbol(symbol.clone());
+        let declarations = {
+            append_if_unique(
+                symbol.maybe_declarations().as_ref().map(|vec| vec.clone()),
+                node,
+            )
+        };
+        symbol.set_declarations(declarations);
+    }
+
+    fn get_declaration_name(&self, node: Rc<Node>) -> Option<__String> {
+        let name = get_name_of_declaration(node);
+        if let Some(name) = name {
+            return if is_property_name_literal(&*name) {
+                Some(get_escaped_text_of_identifier_or_literal(name.clone()))
+            } else {
+                None
+            };
+        }
```

Comment:
This looks like it could be expressed with `and_then` (flat map) instead of an if-statement, but maybe isn't compatible with whatever will replace `unimplemented!`.

Reply:
Does a `return` inside a closure body return from the enclosing function (like a Ruby block) or just from the closure (relevant here because this block that would go inside `.and_then()` needs to return from the function)?

Reply:
Just from the closure